### PR TITLE
CASSANDRA-19923: Add transport extension for coordinated write

### PR DESCRIPTION
**Pull request by yifan-c**

### CODE REVIEW

**Comment by yifan-c** [comment]
After #80

File: scripts/build-sidecar.sh [hunk 1]
```diff
@@ -24,7 +24,7 @@ else
   SCRIPT_DIR=$( dirname -- "$( readlink -f -- "$0"; )"; )
   SIDECAR_REPO="${SIDECAR_REPO:-https://github.com/apache/cassandra-sidecar.git}"
   SIDECAR_BRANCH="${SIDECAR_BRANCH:-trunk}"
-  SIDECAR_COMMIT="${SIDECAR_COMMIT:-55a9efee30555d3645680c6524043a6c9bc1194b}"
+  SIDECAR_COMMIT="${SIDECAR_COMMIT:-f07e248d0ce8303a06daf93b462190ef7be7304d}"
```
Comment (frankgh) [inline scripts/build-sidecar.sh:27]: any reason not to move to the current head of Sidecar? `7f8db256377ff4e449a83282b2561ce5e7b74adf`

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/transports/storage/extensions/StorageTransportHandler.java [hunk 2]
```diff
@@ -89,6 +106,20 @@ private void updateCredentials(UUID jobId, StorageCredentialPair credentialPair)
...
+    {
+        UpdateRestoreJobRequestPayload requestPayload = new UpdateRestoreJobRequestPayload(null, null, status, null);
+        try
+        {
+            // TODO: send to all clusters\
+            transportContext.dataTransferApi().updateRestoreJob(requestPayload);
+        }
+        catch (ClientException e)
+        {
+            throw new RuntimeException("Failed to update secretes for restore job. restoreJobId: " + jobId, e);
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/transports/storage/extensions/StorageTransportHandler.java]: wrong exception verbiage here?

Comment (frankgh) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/transports/storage/extensions/StorageTransportHandler.java]: also you might want to check that the secretes is replaced with secrets

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/transports/storage/extensions/CoordinatedTransportExtension.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+ * The coordinated write has 2 phases, i.e. staging phase and importing phase. In the happy path, the steps of a run are the following:
+ * <ol>
+ *     <li>Extension sets the {@link CoordinationSignalListener} on initialization.</li>
+ *     <li>Extension invokes {@link CoordinationSignalListener#onStageReady(String)},
+ *     when it decides it is time to stage SSTables on all clusters.</li>
+ *     <li>Cassandra Analytics calls Sidecars to stage data.
+ *     {@link #onStageSucceeded(String, long, long, long)} is called per cluster to notify the extension.</li>
+ *     <li>Extension invokes {@link CoordinationSignalListener#onApplyReady(String)},
+ *     when it decides it is time to apply/import SSTables on all clusters.</li>
+ *     <li>ECassandra Analytics calls Sidecars to import data.
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/transports/storage/extensions/CoordinatedTransportExtension.java]:
```suggestion
 *     <li>Cassandra Analytics calls Sidecars to import data.
```

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/transports/storage/extensions/CoordinationSignalListener.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.cassandra.spark.transports.storage.extensions;
+
+/**
+ * A listener interface that receives coordination signals.
+ * It works in cooperation with {@link CoordinatedTransportExtension} to enable coordinated write
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/transports/storage/extensions/CoordinationSignalListener.java:24]:
```suggestion
 * It works in cooperation with {@link CoordinatedTransportExtension} to enable coordinated writes
```

File: cassandra-analytics-core-example/src/main/java/org/apache/cassandra/spark/example/LocalStorageTransportExtension.java [hunk 5]
```diff
@@ -114,4 +116,35 @@ private StorageCredentialPair generateTokens()
...
+    @Override
+    public void onStageSucceeded(String clusterId, long objectsCount, long rowsCount, long elapsedMillis)
+    {
+        LOGGER.info("Job {} has all objects staged at cluster {} after {}ms", jobId, clusterId, elapsedMillis);
+    }
+
+    @Override
+    public void onStageFailed(String clusterId, Throwable cause)
+    {
+        LOGGER.error("Cluster {} failed to stage objects", clusterId, cause);
```
Comment (frankgh) [inline cassandra-analytics-core-example/src/main/java/org/apache/cassandra/spark/example/LocalStorageTransportExtension.java]: nit also include jobId? even though this is a sample class

**Review by frankgh — COMMENTED** [review]
Looks good in general. I've added some feedback

File: scripts/build-sidecar.sh (hunk 1 above)
Comment (yifan-c) [reply scripts/build-sidecar.sh:27]: Thanks. I was not aware that there is a new commit.

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/transports/storage/extensions/CoordinationSignalListener.java (hunk 4 above)
Comment (yifan-c) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/transports/storage/extensions/CoordinationSignalListener.java:24]: the feature name is "coordinated write"

**Review by frankgh — APPROVED** [review]
+1 thanks for addressing the feedback

**Pull request merged**